Vladislav Golovkin/0/main.rb

```diff
@@ -0,0 +1,183 @@
+require 'roo'
+require 'roo-xls'
+
+# application constants
+FIRST_SHEET = 0
+NAME_COL = 0
+AREA_COL = 6
+DATE_ROW = 3
+TABLE_HEADER = 8
+DENOM_BYN = 10_000
+DENOM_YEAR = 2017
+DELTA = 0.2
+DATA_FOLDER = './data/'.freeze
+
+# product properties
+class ProductData
+  attr_accessor :date, :price
+
+  def initialize(price = 0, date = Date.new)
+    @price = price
+    @date = date
+  end
+end
+
+# collection of products
+class ProductsAnalyzer
+  attr_accessor :latest_date, :latest_dictionary
+
+  # @dictionary contains all data from tables
+  # @latest dictionary contains data from the last month
+  def initialize(dictionary = Hash.new([]), latest_dictionary = Hash.new(ProductData), latest_date = Date.new)
+    @dictionary = dictionary
+    @latest_dictionary = latest_dictionary
+    @latest_date = latest_date
+
+    @months = { 'январь' => 1, 'февраль' => 2, 'март' => 3, 'апрель' => 4,
+                'май' => 5, 'июнь' => 6, 'июль' => 7, 'август' => 8,
+                'сентябрь' => 9, 'октябрь' => 10, 'ноябрь' => 11,
+                'декабрь' => 12 }
+  end
+
+  # parses date from table row
+  def get_date(string)
+    values = string.to_s.chomp.split(' ')
+
+    year = nil
+    month = nil
+
+    values.each do |token|
+      month = @months[token] if @months.include? token.downcase
+      year = token.to_i if token =~ /^[-+]?[1-9]([0-9]*)?$/
+    end
+
+    # use 1st day of month for initializing
+    date = Date.new(year, month, 1)
+    @latest_date = date if latest_date < date
+    date
+  end
+
+  # clears latest info, in case of obsoleteness
+  def update_latest_product(latest)
+    @latest_dictionary = Hash.new(ProductData) if latest
+  end
+
+  # adds product to all-data dictionary. Optionally adds product to latest data dictionary
+  def add_entry(product_name, date, price, latest)
+    new_entry = ProductData.new(price, date)
+
+    new_entry.price /= DENOM_BYN if date.year < DENOM_YEAR
```

Comment:
u can write letters for free. use full understandable names --> DENOMINATION_YEAR, DENOMINATION_RATE